src/tophu/label.py

```diff
@@ -0,0 +1,287 @@
+import itertools
+from typing import Set, Tuple
+
+import numpy as np
+from numpy.typing import ArrayLike, NDArray
+from scipy import ndimage
+
+from .tile import TiledPartition
+from .union_find import DisjointSetForest
+
+__all__ = [
+    "deduplicate_labels",
+    "merge_equivalent_labels",
+    "unique_nonzero_integers",
+]
+
+
+NDSlice = Tuple[slice, ...]
+
+
+def deduplicate_labels(
+    conncomp: NDArray[np.unsignedinteger],
+    tiles: TiledPartition,
+) -> None:
+    """
+    De-duplicate connected component labels from different tiles.
+
+    Connected component labels within each tile are updated such that no tile has any
+    label in common with any other tile.
+
+    Only nonzero labels are modified. Zero-valued pixels are treated as masked out (i.e.
+    not part of any connected component).
+
+    The `conncomp` array is modified in-place.
+
+    Parameters
+    ----------
+    conncomp : numpy.ndarray
+        The input connected component labels.
+    tiles : TiledPartition
+        A set of tiles that spans the `conncomp` array. Each tile is represented by an
+        index expression (i.e. a tuple of `slice` objects) that can be used to access
+        a corresponding block of data from the partitioned array.
+
+    See Also
+    --------
+    merge_equivalent_labels
+    """
+    # Keep track of the largest connected component label encountered while looping over
+    # tiles.
+    max_label = 0
+    for tile in tiles:
+        # Get a mask of elements within the tile with nonzero connected component
+        # labels.
+        mask = conncomp[tile] != 0
+
+        # Update connected component labels by adding the largest label value from among
+        # all previously encountered tiles. Zero-valued labels are not changed. The
+        # input connected component labels array is modified in-place.
+        conncomp[tile][mask] += max_label
+
+        # Get the new maximum connected component label.
+        max_label = max(max_label, np.max(conncomp[tile]))
+
+
+def expand_tile(tile: NDSlice, margin: int) -> NDSlice:
+    """
+    Expand a tile by adding some margin to each border.
+
+    Parameters
+    ----------
+    tile : tuple of slice
+        A tuple of slice objects representing a multi-dimensional indexing expression,
+        such as that returned by `numpy.index_exp`.
+    margin : int
+        The number of samples to add to each border of the tile.
+
+    Returns
+    -------
+    expanded_tile : tuple of slice
+        The expanded indexing expression.
+    """
+    if margin < 0:
+        raise ValueError("margin must be >= 0")
+
+    def expand_slice(s: slice) -> slice:
```

Comment:
Thanks for the review, @scottstanie! I appreciate your suggestions and questions, they are very helpful & insightful :smile:

>Is there a difference having this function as a nested function inside `expand_tile`, as opposed to a module-level private function `_expand_slice`? Just trying to clarify that it's only going to be used by this outer function?

Yeah, as you noted, the functional difference is just that the scope of `expand_slice()`  is limited in this case so that it can't be invoked from outside of the function that wraps it.

In general, I find this pattern useful in cases such as: **(A)** I want to write a small helper function that relies on some assumptions that aren't necessarily true in general but are known to be true in the specific case where I'm calling the function, or **(B)** I want to write a small helper function but don't want to write detailed documentation for it. If I'm exposing a function at the module level, even if it's private, I usually try to make sure it has a meaningful name and a detailed docstring. But those things are less important when the function has very limited scope and its behavior can be easily inferred from local context.

In other words, if I'm implementing a function whose implementation is small/simple, but whose preconditions and/or post-conditions are very specific to the context, or are just inconvenient to explain in prose in the docstring, I often find it helpful to limit the scope of that function in order to reduce the potential for misunderstanding or misuse.

This particular case is an example of category **B** -- I don't think there's any reason that `expand_slice()` couldn't be exposed at the module scope, but I couldn't be bothered to give it a docstring. We could pull it out of the wrapper function if you (or anyone) have a strong preference or if we find that we have additional use for it elsewhere later on.

I don't really have a hard rule for determining whether a helper function like this should be function-local or exposed at the module level. I usually just make an instinctive decision based on how specific the use case for the function is, how difficult it is to give it an appropriate name and define its interface, and how annoying it would be to write a detailed docstring for it.

I admit that this practice doesn't seem to be too common in Python programming, so maybe I should cut it out. But, in my mind at least, it serves a useful purpose to limit the scope of certain functions this way.

Reply:
Thanks for explaining, both those cases make sense. It doesn't seem like bad Python style, even if it's less common. At times it seems nicer than a once-used module-level private function so I think it's good here